File: nothing/diagnostics.py
"""Device diagnostics: logcat dump, bugreport, ANR/tombstone collection."""

import datetime
from pathlib import Path

from .device import adb_pull, run
from .exceptions import AdbError
from .models import DeviceInfo

_TEMP = "/data/local/tmp"


# ---------------------------------------------------------------------------
# Logcat
# ---------------------------------------------------------------------------

def action_logcat(
    device: DeviceInfo,
    base_dir: Path,
    package: str | None = None,
    tag: str | None = None,
    level: str | None = None,
    lines: int = 500,
) -> None:
    """
    Dump the current logcat buffer to a local file with optional filters.

    package: filter by app package name (resolved to PID)
    tag:     filter by log tag
    level:   minimum log level — V / D / I / W / E / F (default: all)
    lines:   max lines to capture (default 500)
    """
    # Build filter spec: "tag:level" or "*:V" for all
    if tag and level:
        filter_spec = f"{tag}:{level.upper()} *:S"
    elif tag:
        filter_spec = f"{tag}:V *:S"
    elif level:
        filter_spec = f"*:{level.upper()}"
    else:
        filter_spec = "*:V"

    # Resolve package → PID for --package filter
    pid_filter = ""
    if package:
        r = run(["adb", "-s", device.serial, "shell", f"pidof {package}"])
        pid = r.stdout.strip().split()[0] if r.stdout.strip() else None
        if pid:
            pid_filter = f"--pid={pid}"
            print(f"  Package {package} → PID {pid}")
        else:
            print(f"  [WARN] {package} not running — capturing full buffer")

    ts        = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    dest_dir  = base_dir / "logs"
    dest_dir.mkdir(parents=True, exist_ok=True)

    label = package or tag or "full"
    dest  = dest_dir / f"logcat_{label}_{ts}.txt"

    # -d = dump buffer and exit, -v threadtime = full timestamp + thread info
    cmd = ["adb", "-s", device.serial, "logcat",
           "-d", "-v", "threadtime", "-t", str(lines)]
    if pid_filter:
        cmd.append(pid_filter)
    cmd.append(filter_spec)

    print(f"  Capturing logcat (max {lines} lines, filter: {filter_spec})...")
    r = run(cmd)

    if not r.stdout.strip():
        print("  [WARN] Empty logcat — buffer may have been cleared.")
        return

    dest.write_text(r.stdout, encoding="utf-8", errors="replace")
    line_count = r.stdout.count("\n")
    print(f"[OK] {line_count} lines → {dest}")


# ---------------------------------------------------------------------------
# Bugreport
# ---------------------------------------------------------------------------

def action_bugreport(device: DeviceInfo, base_dir: Path) -> None:
    """
    Trigger adb bugreport and save the ZIP to base_dir/bugreports/.
    Bugreport includes logcat, ANR traces, tombstones, dumpsys, and more.
    This typically takes 30-90 seconds.
    """
    ts       = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    dest_dir = base_dir / "bugreports"
    dest_dir.mkdir(parents=True, exist_ok=True)
    dest     = dest_dir / f"bugreport_{device.serial}_{ts}.zip"

    print(f"  Generating bugreport (this takes 30–90 seconds)...")
    print(f"  Saving to: {dest}")

    # adb bugreport <path> saves directly to the local path
    r = run(["adb", "-s", device.serial, "bugreport", str(dest)])

    if not dest.exists():
        raise AdbError(
            f"Bugreport not created.\n"
            f"Output: {(r.stdout + r.stderr).strip()}"
        )

    size_mb = dest.stat().st_size / 1024 / 1024
    print(f"[OK] Bugreport saved — {size_mb:.1f} MB → {dest}")
    print(f"     Open the ZIP with any archive manager or 'unzip -l \"{dest}\"' on macOS/Linux")


# ---------------------------------------------------------------------------
# ANR / tombstone dump
# ---------------------------------------------------------------------------

def action_anr_dump(device: DeviceInfo, base_dir: Path) -> None:
    """
    Pull ANR traces from /data/anr/ and tombstones from /data/tombstones/.
    Both require root. Saves to base_dir/diagnostics/<timestamp>/.
    """
    ts      = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    dest    = base_dir / "diagnostics" / ts
    dest.mkdir(parents=True, exist_ok=True)

    sources = [
        ("/data/anr",        "anr"),
        ("/data/tombstones", "tombstones"),
    ]

    any_found = False

    for remote_dir, label in sources:
        # Check if directory has any files
        r = run(["adb", "-s", device.serial, "shell",
                 f"su -c 'ls {remote_dir}/ 2>/dev/null | wc -l'"])
        count_str = r.stdout.strip()
        try:
            count = int(count_str)
        except ValueError:
            count = 0

        if count == 0:
            print(f"  {label:<12}: empty (no crashes recorded)")
            continue

        print(f"  {label:<12}: {count} file(s) — copying...")

        # Copy to a readable temp location, then pull
        tmp = f"{_TEMP}/{label}_dump_{ts}"
        r2 = run(["adb", "-s", device.serial, "shell",
                  f"su -c 'cp -r {remote_dir} {tmp} && chmod -R 644 {tmp}/* 2>/dev/null && echo __OK__'"])

        if "__OK__" not in r2.stdout:
            print(f"  [WARN] Could not copy {remote_dir} (root needed?)")
            continue

        local_dest = dest / label
        local_dest.mkdir(exist_ok=True)
        try:
            adb_pull(tmp, local_dest, device.serial)
            any_found = True
            print(f"         saved → {local_dest}/")
        except Exception as e:
            print(f"  [WARN] Pull failed: {e}")
        finally:
            run(["adb", "-s", device.serial, "shell",
                 f"su -c 'rm -rf {tmp}'"])

    if any_found:
        print(f"\n[OK] Diagnostics saved → {dest}")
    else:
        print(f"\n[OK] No ANR traces or tombstones found — device is clean.")
        dest.rmdir()

File: go/internal/diagnostics/diagnostics.go
// Package diagnostics provides logcat, bugreport, and ANR/tombstone collection.
package diagnostics

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

const remoteTmp = "/data/local/tmp"

func timestamp() string {
	return time.Now().Format("20060102_150405")
}

// ActionLogcat dumps the current logcat buffer to a local file with optional filters.
func ActionLogcat(serial, baseDir, packageName, tag, level string, lines int) error {
	// Build filter spec
	filterSpec := "*:V"
	if tag != "" && level != "" {
		filterSpec = tag + ":" + strings.ToUpper(level) + " *:S"
	} else if tag != "" {
		filterSpec = tag + ":V *:S"
	} else if level != "" {
		filterSpec = "*:" + strings.ToUpper(level)
	}

	pidFilter := ""
	if packageName != "" {
		pidOut, _, _ := adb.Run([]string{"adb", "-s", serial, "shell", "pidof " + packageName})
		pidParts := strings.Fields(strings.TrimSpace(pidOut))
		if len(pidParts) > 0 {
			pidFilter = "--pid=" + pidParts[0]
			fmt.Printf("  Package %s \u2192 PID %s\n", packageName, pidParts[0])
		} else {
			fmt.Printf("  [WARN] %s not running \u2014 capturing full buffer\n", packageName)
		}
	}

	ts := timestamp()
	destDir := filepath.Join(baseDir, "logs")
	if err := os.MkdirAll(destDir, 0755); err != nil {
		return err
	}

	label := packageName
	if label == "" {
		label = tag
	}
	if label == "" {
		label = "full"
	}
	dest := filepath.Join(destDir, fmt.Sprintf("logcat_%s_%s.txt", label, ts))

	cmd := []string{"adb", "-s", serial, "logcat", "-d", "-v", "threadtime", "-t", fmt.Sprintf("%d", lines)}
	if pidFilter != "" {
		cmd = append(cmd, pidFilter)
	}
	cmd = append(cmd, filterSpec)

	fmt.Printf("  Capturing logcat (max %d lines, filter: %s)...\n", lines, filterSpec)
	stdout, _, _ := adb.Run(cmd)

	if strings.TrimSpace(stdout) == "" {
		fmt.Println("  [WARN] Empty logcat \u2014 buffer may have been cleared.")
		return nil
	}

	if err := os.WriteFile(dest, []byte(stdout), 0644); err != nil {
		return err
	}
	lineCount := strings.Count(stdout, "\n")
	fmt.Printf("[OK] %d lines \u2192 %s\n", lineCount, dest)
	return nil
}

// ActionBugreport triggers adb bugreport and saves the ZIP to baseDir/bugreports/.
func ActionBugreport(serial, baseDir string) error {
	ts := timestamp()
	destDir := filepath.Join(baseDir, "bugreports")
	if err := os.MkdirAll(destDir, 0755); err != nil {
		return err
	}
	dest := filepath.Join(destDir, fmt.Sprintf("bugreport_%s_%s.zip", serial, ts))

	fmt.Println("  Generating bugreport (this takes 30\u201390 seconds)...")
	fmt.Printf("  Saving to: %s\n", dest)

	stdout, stderr, code := adb.Run([]string{"adb", "-s", serial, "bugreport", dest})

	if _, err := os.Stat(dest); os.IsNotExist(err) {
		return nterrors.AdbError(fmt.Sprintf("Bugreport not created.\nOutput: %s",
			strings.TrimSpace(stdout+stderr)))
	}
	_ = code
	info, err := os.Stat(dest)
	if err != nil {
		return err
	}
	sizeMB := float64(info.Size()) / 1024 / 1024
	fmt.Printf("[OK] Bugreport saved \u2014 %.1f MB \u2192 %s\n", sizeMB, dest)
	fmt.Printf("     Open the ZIP with any archive manager or 'unzip -l \"%s\"' on macOS/Linux\n", dest)
	return nil
}

// ActionANRDump pulls ANR traces and tombstones from the device (requires root).
func ActionANRDump(serial, baseDir string) error {
	ts := timestamp()
	dest := filepath.Join(baseDir, "diagnostics", ts)
	if err := os.MkdirAll(dest, 0755); err != nil {
		return err
	}

	sources := []struct{ path, label string }{
		{"/data/anr", "anr"},
		{"/data/tombstones", "tombstones"},
	}

	anyFound := false

	for _, src := range sources {
		countOut, _, _ := adb.Run([]string{"adb", "-s", serial, "shell",
			fmt.Sprintf("su -c 'ls %s/ 2>/dev/null | wc -l'", src.path)})
		var count int
		fmt.Sscanf(strings.TrimSpace(countOut), "%d", &count)

		if count == 0 {
			fmt.Printf("  %-12s: empty (no crashes recorded)\n", src.label)
			continue
		}
		fmt.Printf("  %-12s: %d file(s) \u2014 copying...\n", src.label, count)

		tmp := fmt.Sprintf("%s/%s_dump_%s", remoteTmp, src.label, ts)
		r2out, _, _ := adb.Run([]string{"adb", "-s", serial, "shell",
			fmt.Sprintf("su -c 'cp -r %s %s && chmod -R 644 %s/* 2>/dev/null && echo __OK__'",
				src.path, tmp, tmp)})

		if !strings.Contains(r2out, "__OK__") {
			fmt.Printf("  [WARN] Could not copy %s (root needed?)\n", src.path)
			continue
		}

		localDest := filepath.Join(dest, src.label)
		if err := os.MkdirAll(localDest, 0755); err != nil {
			fmt.Printf("  [WARN] mkdir failed: %v\n", err)
			continue
		}

		if err := adb.AdbPull(serial, tmp, localDest); err != nil {
			fmt.Printf("  [WARN] Pull failed: %v\n", err)
		} else {
			anyFound = true
			fmt.Printf("         saved \u2192 %s/\n", localDest)
		}
		adb.Run([]string{"adb", "-s", serial, "shell", "su -c 'rm -rf " + tmp + "'"})
	}

	if anyFound {
		fmt.Printf("\n[OK] Diagnostics saved \u2192 %s\n", dest)
	} else {
		fmt.Println("\n[OK] No ANR traces or tombstones found \u2014 device is clean.")
		os.Remove(dest)
	}
	return nil
}
